### PR TITLE
feat(sqllab): SPA migration

### PR DESCRIPTION
**Pull request by justinpark**
### SUMMARY
Part 4-9 of SQLLab SPA migration work: [detail link](https://docs.google.com/document/d/1opyNuaw-CEAD9tG6fx_zqNcZwe3_soQgsnSJTP4mQkk/edit?usp=sharing)

- Change the url from /superset/sqllab to /sqllab
- Update frontend routes
- Update unit tests and cypress e2e tests
- Update redirectSQLLab logic

For the redirect logic, it will replace the SupersetClient.post redirection by react-router redirection with the location state for the post payload data. This means the redundant form_data serialize/deserialize logic has been discarded (which is also another performance gain). 

~~FYI: This commit does *not* include the delete work (i.e. old entrypoint related files, redirect legacy endpoint to new endpoint). This work will be handled in the next PR to minimize the regression rollback.~~ 

### TESTING INSTRUCTIONS

1. Make sure the SQL Lab page works as before but in the context of SPA (does not reload the page when navigating)
2. Make sure that /superset/sqllab redirects to /sqllab including additional query parameters

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by eschutho** [comment]
FYI: This commit does not include the delete work (i.e. old entrypoint related files, redirect legacy endpoint to new endpoint). This work will be handled in the next PR to minimize the regression rollback.
@justinpark are you removing any old api endpoints in the next PR? If so we may want to wait until 4.0. I think there may be some integrations outside of the application that use it, so it could be a breaking change. cc @betodealmeida for the CLI that I think uses this.

**Comment by justinpark** [comment]
> FYI: This commit does not include the delete work (i.e. old entrypoint related files, redirect legacy endpoint to new endpoint). This work will be handled in the next PR to minimize the regression rollback. @justinpark are you removing any old api endpoints in the next PR? If so we may want to wait until 4.0. I think there may be some integrations outside of the application that use it, so it could be a breaking change. cc @betodealmeida for the CLI that I think uses this.

After discussed with other folk, I decided to include the delete work in the same PR. I will post the commit for the delete work soon.

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by michael-s-molina** [comment]
Hi @eschutho. Just to add a little more context. I asked @justinpark to include the cleanup of old code in the same PR given that:
- Some features might work because of old code and they shouldn't. We should validate SPA considering that the old code has no influence in the project.
- The PR is big and requires a lot of testing and validation. It's good if we do this only once. If we do the cleanup work later, we'll need to repeat the same extensive test process.

Just to be clear, removing old code shouldn't mean introducing breaking changes. Let's make sure that's the case when reviewing the PR. 

Once code review is completed, we'll create a test environment to make sure everything is working. It would be great if @sadpandajoe and @jinghua-qa could help us out.

File: superset/models/core.py
Comment (michael-s-molina) [inline superset/models/core.py:497]: Do we need to preserve the old `/superset/sqllab`?

File: tests/integration_tests/core_tests.py
Comment (michael-s-molina) [inline tests/integration_tests/core_tests.py:1162]: Nice!

File: tests/integration_tests/sqllab_tests.py
Comment (michael-s-molina) [inline tests/integration_tests/sqllab_tests.py:270]: Can we move the history to `/sqllab/history` given that is under SPA?

**Review by michael-s-molina — COMMENTED** [review]
Thank you for the PR @justinpark. I left some first-pass comments.

File: superset/models/core.py
Comment (justinpark) [reply superset/models/core.py:497]: Good catch

**Comment by justinpark** [comment]
@michael-s-molina I just changed all addressed issues.

**Comment by kgabryje** [comment]
Code looks good! Great to see the old obsolete code gone

**Comment by eschutho** [comment]
@justinpark and @michael-s-molina This all looks good to me. I didn't see any breaking changes in the PR, as the current endpoints have all the redirects in place as @justinpark mentioned. It's only the deleting of the endpoints that would raise any concerns for breaking changes. 👍

**Review by michael-s-molina — APPROVED** [review]
Thank you @eschutho and @kgabryje. Code LGTM too.

@sadpandajoe @jinghua-qa Could you also review the PR before we merge it?

**Comment by justinpark** [comment]
@sadpandajoe @jinghua-qa Could you help QA test for SPA migration work?

**Comment by sadpandajoe** [comment]
> @sadpandajoe @jinghua-qa Could you help QA test for SPA migration work?

Sure @justinpark. Bring up an ephemeral. Also what have you tested so far so we don't duplicate testing.

**Comment by justinpark** [comment]
> Sure justinpark. Bring up an ephemeral. Also what have you tested so far so we don't duplicate testing.

@sadpandajoe Ephemeral environment already set up at http://34.219.198.150:8080/. 

We completed the basic page transition (from chart to sqllab. etc) so I hope you test running queries and make a bunch of page transition and page reload that make sure no data lost or error thrown.

**Comment by justinpark** [comment]
@sadpandajoe Do you have any updates in your side? 
We're planning to merge it this Wednesday if you don't have any inputs by then.

**Comment by sadpandajoe** [comment]
> @sadpandajoe Do you have any updates in your side? We're planning to merge it this Wednesday if you don't have any inputs by then.

@justinpark I haven't had a chance to really look at it yet since I got stuck doing other things. I think things I was going to look at for testing include:

- Saved queries links still work
- The different navigations back to sql lab like on a chart where you can view query in sql lab
- Stopping queries that are running

But probably won't have time to get to it until the later half of this week as I need to finish some other stuff. I would say merge and if I find anything I'll just create new issues and ping you on them.

**Comment by eschutho** [comment]
I did a quick test on the ephemeral, and it's very snappy! 🙌

**Comment by justinpark** [comment]
> But probably won't have time to get to it until the later half of this week as I need to finish some other stuff. I would say merge and if I find anything I'll just create new issues and ping you on them.

Sounds good! We've completed reasonable tests so I'll take cake in the following bug fixes for your reported bugs.

**Comment by geido** [comment]
@justinpark code LGTM. 

Not sure if all of these are related, actually this might all be unrelated, but for reference here is what I found from manual testing and messing up with SQL Lab

1. The duplicate tab does not recognize tab name and goes with "Untitled". Also, when going back to previous tab the result set is emptied

https://github.com/apache/superset/assets/60598000/26d873f1-54cb-452e-8415-682eab80f4cb

2. When from saved queries to SQL Lab but the menu is still highlighting Saved Queries. Same happens with query history

Screenshot 2023-10-03 at 14 34 13

3. Again, this is probably unrelated but when I have multiple tabs, if I select "Query History" in one tab it will move all the other tabs to "Query History". I think this is related to the video above. For instance, if I am viewing the definition of a table and then switch to a different query it will go empty and it will make the table definition disappear in the originating tab

https://github.com/apache/superset/assets/60598000/f8f908cc-f927-40bd-878e-c83a27d0caae

4. Not sure how this is happening but I found queries in Query history with no title, which will correspond to a tab named "Copy of" in SQL Lab.

5. Found that you can name a query with an empty space (100% unrelated)

6. I saw that we lost the ability to forcefully open the query in a new browser tab from Explore -> Run in SQL Lab. I think the user should still be able to choose whether to open it in a new browser tab or not. Same happens in Saved Queries but not in Query History

**Comment by michael-s-molina** [comment]
Thank you for all the tests and comments @geido! 

@justinpark Given that this PR contains 54 changed files, I highly recommend we fix the bugs that are related to this PR, merge the PR, and then open follow-ups for bugs that already exist on master. The reason for this suggestion is to facilitate the review process given the size of the PR.

**Comment by justinpark** [comment]
> Not sure if all of these are related, actually this might all be unrelated, but for reference here is what I found from manual testing and messing up with SQL Lab
> 
> 1. The duplicate tab does not recognize tab name and goes with "Untitled". Also, when going back to previous tab the result set is emptied

This is an existing bug and will follow up in the separate PR.
 
>  Superset.mp4 
> 2. Wnet from saved queries to SQL Lab but the menu is still highlighting Saved Queries. Same happens with query history
> 
> Screenshot 2023-10-03 at 14 34 13

I think you navigate back/forth by the browser navigation (not menu). This is same issue on existing tabs (went from Saved Queries to Query History) and will follow up in the separate PR.


> 3. Again, this is probably unrelated but when I have multiple tabs, if I select "Query History" in one tab it will move all the other tabs to "Query History". I think this is related to the video above. For instance, if I am viewing the definition of a table and then switch to a different query it will go empty and it will make the table definition disappear in the originating tab
> 

This is how SQL Lab designed before (because activeSouthPaneTab not belong to each tab but belong to the entire sqllab). Should follow up in the separate PR.

> 4. I found queries in Query history with no title, which will correspond to a tab named "Copy of" in SQL Lab.

Same as no. 1

> 5. Realized that you can name a query with an empty space (100% unrelated)

This is an existing bug and will follow up in the separate PR.

> 6. I saw that we lost the ability to forcefully open the query in a new browser tab from Explore -> Run in SQL Lab. I think the user should still be able to choose whether to open it in a new browser tab or not. Same happens in Saved Queries but not in Query History

The basically it should support cmd + click that can open the link in a new browser tab but some links(which you mentioned) require a post data to retrieve the sql context so hard to maintain in SPA approach. I hope this is not a blocker of running SPA and want to discuss in the next thread(PR) for the support of opening new tab.

**Comment by justinpark** [comment]
@rusackas @eschutho @geido  could you help to stamp the PR to unblock the reviewer req?

**Review by rusackas — APPROVED** [review]
Approving cypress changes as codeowner

**Comment by rusackas** [comment]
Oh, and THANK YOU for making this huge stride in the SPA effort. This is a big deal, and we're almost at the end of the tunnel! :D

**Pull request merged**
